routers/web/user/auth/webauthn.go

```diff
@@ -0,0 +1,161 @@
+		return
+	}
+
+	userVerification := ctx.FormString("user_ver")
+	txAuthExtension := ctx.FormString("test_extension")
+	testExtension := protocol.AuthenticationExtensions(map[string]interface{}{"txAuthSimple": txAuthExtension})
+
+	assertion, sessionData, err := wa.WebAuthn.BeginLogin(
+		(*wa.User)(user),
+		webauthn.WithUserVerification(protocol.UserVerificationRequirement(userVerification)),
+		webauthn.WithAssertionExtensions(testExtension),
+	)
+	if err != nil {
+		ctx.ServerError("webauthn.BeginLogin", err)
+		return
+	}
+
+	if err := ctx.Session.Set("assertion", sessionData); err != nil {
+		ctx.ServerError("Session.Set", err)
+		return
+	}
+	ctx.JSON(http.StatusOK, assertion)
+}
+
+// WebAuthnLoginAssertionPost validates the signature and logs the user in
+func WebAuthnLoginAssertionPost(ctx *context.Context) {
+	idSess := ctx.Session.Get("twofaUid")
+	sessionData := ctx.Session.Get("assertion")
+	if sessionData == nil || idSess == nil {
+		ctx.ServerError("UserSignIn", errors.New("not in WebAuthn session"))
+		return
+	}
+	user, err := user_model.GetUserByID(idSess.(int64))
+	if err != nil {
+		ctx.ServerError("UserSignIn", err)
+		return
+	}
+
+	log.Trace("Finishing webauthn authentication with user: %s\n", user.Name)
+
+	// With the session data retrieved, we need to call webauthn.FinishLogin to
+	// verify the signed challenge. This returns the webauthn.Credential that
+	// was used to authenticate.
+	cred, err := wa.WebAuthn.FinishLogin((*wa.User)(user), *sessionData.(*webauthn.SessionData), ctx.Req)
+	if err != nil {
+		ctx.ServerError("FinishLogin", err)
+		return
+	}
+
+	// At this point, we've confirmed the correct authenticator has been
+	// provided and it passed the challenge we gave it. We now need to make
+	// sure that the sign counter is higher than what we have stored to help
+	// give assurance that this credential wasn't cloned.
+	if cred.Authenticator.CloneWarning {
+		log.Error("credential appears to be cloned: %s", err)
+		ctx.Status(http.StatusForbidden)
+		return
+	}
+	// We're logged in! All that's left is to update the sign count with the
+	// new value we received. We could join the tables on the CredentialID
+	// field, but for our purposes we'll just get the stored credential and
+	// use that to find the authenticator we need to update.
+	dbCred, err := login.GetWebAuthnCredentialByCredID(base64.RawStdEncoding.EncodeToString(cred.ID))
+	if err != nil {
+		ctx.ServerError("UserSignIn", err)
+	}
+
+	dbCred.SignCount = cred.Authenticator.SignCount
+	if err := dbCred.UpdateSignCount(); err != nil {
+		ctx.ServerError("UserSignIn", err)
+	}
+
+	if ctx.Session.Get("linkAccount") != nil {
+		gothUser := ctx.Session.Get("linkAccountGothUser")
+		if gothUser == nil {
+			ctx.ServerError("UserSignIn", errors.New("not in LinkAccount session"))
+			return
+		}
+
+		err = externalaccount.LinkAccountToUser(user, gothUser.(goth.User))
+		if err != nil {
+			ctx.ServerError("UserSignIn", err)
+			return
+		}
+	}
```

Comment:
IIRC, there are similar code in `2fa.go` and there is a function `LinkAccountFromStore`